nntrainer/graph/network_graph.cpp

```diff
@@ -0,0 +1,424 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2020 Jijoong Moon <jijoong.moon@samsung.com>
+ *
+ * @file    network_graph.h
+ * @date    19 Oct 2020
+ * @see     https://github.com/nnstreamer/nntrainer
+ * @author  Jijoong Moon <jijoong.moon@samsung.com>
+ * @bug     No known bugs except for NYI items
+ * @brief   This is Network Graph Class for Neural Network
+ *
+ */
+
+#include <layer_factory.h>
+#include <network_graph.h>
+
+#include <nntrainer_error.h>
+#include <nntrainer_log.h>
+#include <parse_util.h>
+
+namespace nntrainer {
+
+void NetworkGraph::addEdge(unsigned int ith, LayerNode node) {
+  if (ith > num_node - 1)
+    throw std::invalid_argument("Exceed total number of layer");
+
+  adj[ith].push_back(node);
+}
+
+void NetworkGraph::addLayerNode(std::shared_ptr<Layer> layer) {
+  std::list<LayerNode> l;
+  LayerNode *node = new LayerNode();
+
+  node->layer = layer;
+  node->index = num_node;
+
+  l.assign(1, *node);
+  adj.push_back(l);
+  num_node++;
+}
+
+void NetworkGraph::topologicalSortUtil(unsigned int ith, bool visited[],
+                                       std::stack<LayerNode> &Stack) {
+  visited[ith] = true;
+
+  std::list<LayerNode>::iterator i;
+  for (i = adj[ith].begin(); i != adj[ith].end(); ++i) {
+    if (!visited[(*i).index])
+      topologicalSortUtil((*i).index, visited, Stack);
+  }
+
+  Stack.push(getLayerNode(ith));
+}
+
+LayerNode &NetworkGraph::getLayerNode(unsigned int ith) {
+
+  std::list<LayerNode>::iterator iter;
+  for (unsigned int i = 0; i < num_node; ++i) {
```

Comment:
node->index is made using num_node and its location is at the ith index itself in the `adj` vector.
Lets make this `getLayerNode` as a constant time operation than O(N) operation.

Reply:
We might need to change the order ( not yet supported ). Let's assume that user write loss for the first element in ini file but set properly as an last layer. In that case object might be in first but the index must be the last.

Reply:
as per line 35, index of the node is assigned at the time of node creation and it inserted in the adj list at that moment.
adj list order also remains same as it represent the graph data. So, I dont understand how a node in adj list and order of layer in ini file are related.

Reply:
> as per line 35, index of the node is assigned at the time of node creation and it inserted in the adj vector at that moment.
> adj vector order also remains same as it represent the graph data. So, I dont understand how a node in adj vector and order of layer in ini file are related.
> In my understanding, the order of the node->index is just the order in which nodes are created and represents no other information (same for order of nodes in adj vector)

Yes. you can see it is now. What I'm saying that will be changed soon when we do TODO After make node list of graph, we have to find root. --> When we do find the root then, I think index will be re-ordered. That means idx is not same any more.

Reply:
The index of the nodes remains same as they dont represent any ordering. Root of the graph can be any node and can be multiple (will have to be supported some day). So, the order of nodes in adj list should not have a meaning. If needed, there can be a different list which has meaning to order of nodes like Sorted.

Reply:
> The index of the nodes remains same as they dont represent any ordering. Root of the graph can be any node and can be multiple (will have to be supported some day). So, the order of nodes in adj list should not have a meaning. If needed, there can be a different list which has meaning to order of nodes like Sorted.

Let's leave as it is. we need it now. We will see and if we don't need it , then we remove later.